src/shared/udev-util.c

```diff
@@ -584,59 +581,62 @@ int udev_queue_init(void) {
 }
 
 int on_ac_power(void) {
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
         bool found_offline = false, found_online = false;
-        _cleanup_closedir_ DIR *d = NULL;
+        sd_device *d;
         int r;
 
-        d = opendir("/sys/class/power_supply");
-        if (!d)
-                return errno == ENOENT ? true : -errno;
+        r = sd_device_enumerator_new(&e);
+        if (r < 0)
+                return r;
```

Comment:
Hmm, what are the effects of this change if udevd is not running or not fully initialized?

Reply:
Hm? I thought that `sd_device_enumerator` does not require udevd, especially when `allow_uninitialized()` is called.

Reply:
I think that was the idea when it was introduced in dee5e0b6c29ca22cbf67d28fbee195d70b0dde72. But we don't have any docs and I'm not sure that it still works like this… I think we should verify that it behaves as expected before merging this.

Reply:
After re-reading sd-device code, actually it does not depend on udevd if `sd_device_enumerator` with `sd_device_enumerator_allow_uninitialized()` called, and only `sd_device_get_sysattr_value()` and `sd_device_get_parent()` are called for enumerated devices.

```
$ sudo udevadm info --cleanup-db
$ ls /run/udev/*
/run/udev/control

/run/udev/data:

/run/udev/links:

/run/udev/rules.d:

/run/udev/static_node-tags:

/run/udev/tags:

/run/udev/watch:
$ SYSTEMD_LOG_LEVEL=debug build/systemd-ac-power
sd-device-enumerator: Scan all dirs
sd-device-enumerator: Scanning /sys/bus
sd-device-enumerator: Scanning /sys/class
AC: The power supply is currently online.
BAT0: The power supply is battery, ignoring.
port0: The USB type-C port is in power sink mode.
port1: The USB type-C port is in power sink mode.
ucsi-source-psy-USBC000:001: The USB type-C device has at least one port in power sink mode.
ucsi-source-psy-USBC000:001: The power supply is currently offline.
port0: The USB type-C port is in power sink mode.
port1: The USB type-C port is in power sink mode.
ucsi-source-psy-USBC000:002: The USB type-C device has at least one port in power sink mode.
ucsi-source-psy-USBC000:002: The power supply is currently online.
Found at least one online non-battery power supply, system is running on AC power.
```

Reply:
Note that `sd-device` depends on udevd when,
- sd_device_monitor is monitoring netlink message from udevd.
- sd_device_enumerator accepts only initialized (hence, processed by udevd) devices.
- sd_device object called with sd_device_get_property_value(), sd_device_get_is_initialized(), sd_device_get_usec_initialized(), sd_device_get_tag_first/next(), sd_device_get_current_tag_first/next(), sd_device_get_devlink_first/next(), sd_device_has_tag(), sd_device_has_current_tag(), and so on.

Reply:
That's great info, can we get it documented somewhere please? Maybe docs/architecture.md?

Reply:
Sure, I will do that later.